kafka/coordinator/base.py

```diff
@@ -758,6 +843,112 @@ def __init__(self, heartbeat, metrics, prefix, tags=None):
+        try:
+            while not self.closed:
+                self._run_once()
+
+            log.debug('Heartbeat closed!')
+
+        except RuntimeError as e:
+            log.error("Heartbeat thread for group %s failed due to unexpected error: %s",
+                      self.coordinator.group_id, e)
+            self.failed = e
+
+    def _run_once(self):
+        with self.coordinator._lock:
+            if not self.enabled:
+                log.debug('Heartbeat disabled. Waiting')
+                self.coordinator._lock.wait()
+                log.debug('Heartbeat re-enabled.')
+                return
+
+            if self.coordinator.state is not MemberState.STABLE:
+                # the group is not stable (perhaps because we left the
+                # group or because the coordinator kicked us out), so
+                # disable heartbeats and wait for the main thread to rejoin.
+                log.debug('Group state is not stable, disabling heartbeats')
+                self.disable()
+                return
+
+            # When consumer.wakeup() is implemented, we need to
+            # disable here to prevent propagating an exception
+            # to this heartbeat thread
+            self.coordinator._client.poll(timeout_ms=0)
+
+            if self.coordinator.coordinator_unknown():
+                if not self.coordinator.lookup_coordinator().is_done():
+                    self.coordinator._lock.wait(self.coordinator.config['retry_backoff_ms'] / 1000)
+
+            elif self.coordinator.heartbeat.session_timeout_expired():
+                # the session timeout has expired without seeing a
+                # successful heartbeat, so we should probably make sure
+                # the coordinator is still healthy.
+                log.debug('Heartbeat session expired, marking coordinator dead')
+                self.coordinator.coordinator_dead('Heartbeat session expired')
+
+            elif self.coordinator.heartbeat.poll_timeout_expired():
+                # the poll timeout has expired, which means that the
+                # foreground thread has stalled in between calls to
+                # poll(), so we explicitly leave the group.
+                log.debug('Heartbeat poll expired, leaving group')
+                self.coordinator.maybe_leave_group()
+
+            elif not self.coordinator.heartbeat.should_heartbeat():
+                # poll again after waiting for the retry backoff in case
+                # the heartbeat failed or the coordinator disconnected
+                log.debug('Not ready to heartbeat, waiting')
+                self.coordinator._lock.wait(self.coordinator.config['retry_backoff_ms'] / 1000)
+
+            else:
+                self.coordinator.heartbeat.sent_heartbeat()
+                future = self.coordinator._send_heartbeat_request()
+                future.add_callback(self._handle_heartbeat_success)
+                future.add_errback(self._handle_heartbeat_failure)
```

Comment:
Taken straight from java client. I think I changed 'heartbeat.send_heartbeat()` to `heartbeat.sent_heartbeat()` just to keep myself somewhat sane. But the flow and internal logic are the same.

Reply:
Yea, you can get that from reading Java code. It never shrinks, parts are only added on top. Probably because a lot of people are working on it, but really is hard to follow sometimes.